packages/desmos-std/src/iter/page_iterator.rs

```diff
@@ -0,0 +1,211 @@
+                Ok(page) => {
+                    // No items on the new page so no more items, flag the iterator as consumed
+                    if page.items.is_empty() {
+                        self.consumed = true;
+                        None
+                    } else {
+                        // Clone the first item
+                        let first_item = page.items[0].clone();
+                        // Save the fetched page
+                        self.current_page = Some(page);
+                        // Update the index for the next iteration
+                        self.page_item_index = 1;
+                        // Return the first item of the new fetched page
+                        Some(Ok(first_item))
+                    }
+                }
+                // An error occurred, propagate it to the caller
+                Err(e) => {
+                    // Set the iterator as consumed to prevent other invocations
+                    self.consumed = true;
+                    Some(Err(e))
+                }
+            }
+        } else {
+            // A page is available and we don't have iterated over all the items
+            let page = self.current_page.as_ref().unwrap();
+            let result = Some(Ok(page.items[self.page_item_index].clone()));
+            // Update the iterator index
+            self.page_item_index += 1;
+            result
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::iter::page_iterator::{Page, PageIterator};
+    use cosmwasm_std::StdError;
+
+    #[test]
+    fn test_iterations_without_errors() {
+        let it: PageIterator<u64, u64> = PageIterator::new(
+            Box::new(|key, limit| {
+                let start = key.unwrap_or(0);
+                Ok(Page {
+                    items: (start..start + limit).collect(),
+                    next_page_key: if start + limit >= 20 {
+                        None
+                    } else {
+                        Some(start + limit)
+                    },
+                })
+            }),
+            10,
+        );
+
+        let mut it_counter = 0;
+        for element in it {
+            element.unwrap();
+            it_counter += 1;
+        }
+
+        assert_eq!(20, it_counter);
+    }
+
+    #[test]
+    fn test_iterations_with_error() {
+        let mut it = PageIterator::<i32, i32>::new(
+            Box::new(|_, _| Err(StdError::generic_err("ERROR :("))),
+            10,
+        );
+
+        let item = it.next();
+
+        // First item should be some with the error that happen fetching the page
+        assert!(item.is_some());
+        assert!(item.unwrap().is_err());
+
+        // Since an error occurred when fetching the page the second item should be None
+        // to signal that the iterator don't have more items.
+        let item = it.next();
+        assert!(item.is_none());
+    }
+
+    #[test]
+    fn test_iterations_with_partial_page() {
```

Comment:
Could you and the test with the case having next page?

Reply:
A test where the first page is full and the second one is partial?

Reply:
@manu0466 yeah, it is.

Reply:
Test added